### PR TITLE
feat(#191): cochange-validate benchmark for blast-radius precision/recall

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

No validation exists for whether blast-radius co-change predictions are accurate. This PR introduces the `cochange-validate` benchmark binary that measures prediction accuracy at multiple Jaccard thresholds against real OSS repository commit history, establishing a baseline for threshold tuning.

- New `cochange-validate` binary with parallel repo processing (3-thread cap per DD-3)
- Temporal train/test split (chronological, no leakage) with quality gates (≥50 multi-file commits, ≥6 months span)
- Reports precision/recall at both macro (per-commit) and micro (per-query) averaging across 6 thresholds

## Changes

**New: `crates/rskim-bench/src/cochange/`**
- `types.rs` — `ThresholdMetrics`, `RepoCochangeResult`, `CochangeValidationResult`, `RunMetadata`, `RepoManifest`; all derive serde with roundtrip tests
- `deny_list.rs` — 10 lock-file names, 8 directory prefixes, 4 extension suffixes; `filter_denied()` via `Vec::retain`
- `temporal_split.rs` — reverses newest-first GixSource output to chronological order before split; handles same-timestamp edge case
- `validate.rs` — `build_path_map`, `compute_precision/recall/f1`, `check_quality_gates`, `evaluate_at_thresholds` (uses `jaccard(a,b)` O(log n), never `pairs_for_file`), `validate_repo` orchestrator, `aggregate_metrics`
- `report.rs` — `to_json` and `to_markdown` with threshold table, per-repo breakdown, methodology, and reproducibility manifest

**New: `crates/rskim-bench/src/bin/cochange_validate.rs`**
- clap CLI: `--corpus-dir`, `--corpus-config`, `--format`, `--save-report`, `--thresholds`, `--train-fraction`
- Parallel processing via `rayon::ThreadPoolBuilder::new().num_threads(3)`

**New: `crates/rskim-research/cochange-corpus.toml`**
- 7 pinned OSS repos (ripgrep, fd, flask, pydantic, gin, fiber, nest) with `deep_clone = true`

**Modified: `crates/rskim-research/src/clone.rs`**
- `clone_with_history(url, dest)` — full clone (no `--depth 1`) for history-dependent analysis; idempotent if dest exists; reuses `git_run_with_timeout` and `security_args`

**Modified: `crates/rskim-research/src/config.rs`**
- `RepoEntry.deep_clone: bool` with `#[serde(default)]` — existing corpus.toml files unaffected

## Breaking Changes

None — all changes are purely additive. The `deep_clone` field on `RepoEntry` uses `#[serde(default)]` so existing TOML files without the field continue to work.

## Reviewer Focus Areas

- **Temporal split correctness** (`temporal_split.rs`): GixSource returns newest-first; the function reverses to chronological before splitting. Verify the leakage-prevention logic and edge-case handling (same timestamp, single commit).
- **Evaluation path performance** (`validate.rs:evaluate_at_thresholds`): iterates all (query, candidate) pairs calling `reader.jaccard()` — O(log n) per pair. This is intentional per the plan constraint against `pairs_for_file`.
- **Quality gate thresholds** (`validate.rs`): ≥50 multi-file commits and ≥6 months span. Are these appropriate for the 7 target repos?
- **Deny-list completeness** (`deny_list.rs`): false-positive tests for `clockwork.rs`, `src/lock.rs`, `vendors/`, `distribution.rs` are included.
- **Unmapped file handling**: files in test commits not present in training `path_map` are excluded from recall denominator and tracked in `unmapped_files_in_test`.

### CODE REVIEW

**Comment by dean0x** [comment]
## Code Review Summary

**Overall Status**: CHANGES_REQUESTED (4 blocking issues at ≥85% confidence)

### Blocking Issues (Already Posted as Inline Comments)

All blocking issues are ≥85% confidence and located in changed code:

1. **Deny-list pattern duplication** (Confidence: 95%) - `cochange_validate.rs:238`
   - Two sources of truth will drift; expose `pattern_names()` from `deny_list.rs`

2. **chrono_now timestamp error** (Confidence: 95%) - `cochange_validate.rs:207`
   - Leap year calculation broken; year off by ~1 after 2025; use `date` command or proper library

3. **Path traversal vulnerability** (Confidence: 85%) - `validate.rs:327`
   - Repo name lacks validation; reuse existing `extract_repo_name()` from `clone.rs`

4. **capture_head_sha missing timeout** (Confidence: 92%) - `validate.rs:618`
   - Git subprocess can hang indefinitely; apply `git_run_with_timeout` pattern

### Additional MEDIUM Issues (60-79% Confidence) — Suggested Fixes

**Complexity/Architecture (Should Fix):**
- **validate_repo god function** (`validate.rs:321-494`) - 170 lines, 9 sequential steps. Extract clone-and-parse phase and build-and-evaluate phase into helpers; introduce `try_or_error_result` macro to reduce `error_result` boilerplate.
  
- **Aggregate micro metrics mismatch** (`validate.rs:531-590`) - Currently averaging per-repo micro values (macro-average), not true micro-average. Propagate raw TP/predicted/actual counts through `ThresholdMetrics` for correct aggregation at both levels.

- **evaluate_at_thresholds O(T*C*F*A) complexity** (`validate.rs:215-260`) - Recomputes Jaccard scores for every threshold; restructure to compute scores once per query-file pair, then sweep thresholds. (Confidence: 85%)

**Reliability/Performance (Code You Touched):**
- **build_path_map allocates before deduping** (`validate.rs:52-63`) - Collects all paths into `Vec` then sorts+deduplicates; use `BTreeSet` as PR description states to deduplicate during insertion. (Confidence: 80%)

- **clone_with_history timeout adequacy** (`clone.rs:295-323`) - Full clone (no `--depth 1`) may exceed shared timeout on large repos (pydantic: 20k+ commits); add `--single-branch` to reduce clone size. (Confidence: 80%)

- **temporal_split clones commit vec** (`temporal_split.rs:88-89`) - `commits.to_vec()` + `.reverse()` allocates all `CommitInfo` structs; accept owned data or reverse in-place since caller owns the vec. (Confidence: 82%)

### Review Scores

| Category | Score | Status |
|----------|-------|--------|
| Architecture | 7/10 | CHANGES_REQUESTED |
| Rust Best Practices | 7/10 | CHANGES_REQUESTED |
| Security | 8/10 | CHANGES_REQUESTED (1 HIGH: path traversal) |
| Reliability | 7/10 | CHANGES_REQUESTED (2 HIGH: timeout, validation) |

### Positive Findings (Well Done)

✓ Clean module separation (types/deny_list/temporal_split/validate/report)
✓ Correct layering (bench crate depends on search, not reverse)
✓ Well-documented public APIs
✓ Proper Strategy pattern for output formats
✓ HTTPS-only clone restriction + git security hardening
✓ Bounded thread pool concurrency (3 threads per DD-3)
✓ NaN guards on train_fraction
✓ Graceful error recovery via error_result
✓ Comprehensive test coverage

### Next Steps

1. Fix the 4 blocking inline issues (all in new code, no pre-existing issues suppressed)
2. Address MEDIUM issues per provided fixes
3. Re-run tests and benchmarks
4. Ping for re-review

---

*Review completed by: Architecture Reviewer, Rust Reviewer, Security Reviewer, Reliability Reviewer*
*Confidence filtering applied: Inline comments ≥80%, summary ≥60%*

File: crates/rskim-bench/src/bin/cochange_validate.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            repo_url: r.repo_url.clone(),
+            head_sha: r.head_sha.clone(),
+            train_cutoff_timestamp: r.split_timestamp,
+            train_commit_count: r.train_commits,
+            test_commit_count: r.test_commits,
+        })
+        .collect()
+}
+
+fn deny_list_pattern_names() -> Vec<String> {
```
Comment (dean0x) [inline crates/rskim-bench/src/bin/cochange_validate.rs]: **Deny-list pattern duplication (HIGH CONFIDENCE: 95%)**

The `deny_list_pattern_names()` function manually re-lists 21 patterns already defined as constants in `deny_list.rs`. These two sources of truth will drift—DRY violation.

**Fix:** Expose `pub fn pattern_names()` from `deny_list.rs` reading from existing constants, then call it here instead of duplicating.

Reviewed by: Architecture, Rust

File: crates/rskim-bench/src/bin/cochange_validate.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        anyhow::bail!("--thresholds must contain at least one value");
+    }
+
+    thresholds.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
+    thresholds.dedup_by(|a, b| (*a - *b).abs() < 1e-9);
+
+    Ok(thresholds)
+}
+
+fn chrono_now() -> String {
```
Comment (dean0x) [inline crates/rskim-bench/src/bin/cochange_validate.rs:283]: **chrono_now timestamp calculation is incorrect (HIGH CONFIDENCE: 95%)**

Year calculation ignores leap years (error ~13-14 days by 2026), month/day replaced with XX-XX. This breaks ISO-8601 reproducibility timestamp claim.

**Fix:** Use `date -u +%Y-%m-%dT%H:%M:%SZ` or a proper time library instead of hand-rolling.

Reviewed by: Architecture, Rust, Reliability

File: crates/rskim-bench/src/cochange/validate.rs [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+/// This is the top-level orchestrator that ties together cloning, history
+/// parsing, filtering, quality gating, splitting, matrix building, and
+/// evaluation.
+pub fn validate_repo(
+    entry: &RepoEntry,
+    corpus_dir: &Path,
+    thresholds: &[f64],
+    train_fraction: f64,
+) -> anyhow::Result<RepoCochangeResult> {
+    let repo_name = entry
```
Comment (dean0x) [inline crates/rskim-bench/src/cochange/validate.rs]: **Path traversal vulnerability: repo name lacks validation (HIGH CONFIDENCE: 85%)**

Repo name extracted without validation (skips `extract_repo_name` checks from clone.rs). URL like `https://evil.com/..` produces `repo_name=".."` → `dest` escapes corpus_dir.

**Fix:** Reuse existing `extract_repo_name()` from rskim_research::clone or inline its validation checks.

Reviewed by: Security, Reliability

File: crates/rskim-bench/src/cochange/validate.rs [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+        commits_skipped_too_large: 0,
+        split_timestamp: 0,
+        metrics_by_threshold: vec![],
+        quality_gate_passed: false,
+        quality_gate_reason: None,
+        error: Some(error),
+    }
+}
+
+fn capture_head_sha(repo_path: &Path) -> anyhow::Result<String> {
```
Comment (dean0x) [inline crates/rskim-bench/src/cochange/validate.rs:864]: **capture_head_sha lacks subprocess timeout (HIGH CONFIDENCE: 92%)**

Calls `Command::new("git").output()` without timeout. Hanging git subprocess blocks rayon thread indefinitely. `clone_with_history` correctly uses `git_run_with_timeout`, but this diverges.

**Fix:** Apply `git_run_with_timeout` pattern or implement similar timeout. Set `credential.helper=` to prevent prompts.

Reviewed by: Reliability

**Comment by dean0x** [comment]
**Clippy `cmp_owned` violations (HIGH CONFIDENCE: 95%)**

Allocating `PathBuf` solely for comparison fails Cargo.toml's `-D warnings` policy. Lines 314–315 both call `PathBuf::from(...)` for comparison against `f.path`.

**Fix:** Use `Path` reference or string comparison instead:

```rust
assert!(files.iter().any(|f| f.path.as_path() == Path::new("src/main.rs")));
assert!(files.iter().any(|f| f.path.as_path() == Path::new("src/lib.rs")));
```

This applies ADR-001 (fix noticed issues immediately) — the lint failure prevents CI from passing with `--all-targets --all-features`.

---
*Reviewed by: Rust, DevFlow*

**Comment by dean0x** [comment]
**O(Q × F) jaccard cache causes large per-commit allocations (HIGH CONFIDENCE: 85%)**

Location: `crates/rskim-bench/src/cochange/validate.rs:279`

For each commit, `compute_jaccard_cache` materializes a `Vec<Vec<(FileId, f64)>>` with up to 500 files × 20,000 total files = 10M entries (~120 MB per commit). With 50K test commits, allocator churn is significant.

**Fix:** Pre-allocate once outside the commit loop and reuse via `clear()` + refill, like the existing `predicted_scratch` pattern.

---
*Reviewed by: Performance*

**Comment by dean0x** [comment]
**`sweep_thresholds` 12 parameters exceeds reasonable complexity (HIGH CONFIDENCE: 85%)**

Location: `crates/rskim-bench/src/cochange/validate.rs:702`

The function signature accepts 4 data inputs + 1 scratch buffer + 7 mutable accumulator slices. While `#[allow(clippy::too_many_arguments)]` acknowledges this, the 7 accumulators form a cohesive concept that should be encapsulated as a struct to prevent index-alignment bugs.

**Fix:** Extract `ThresholdAccumulators` struct containing the 7 accumulator vecs, with methods:
- `fn new(n: usize) -> Self`
- `fn accumulate_query(...)`
- `fn finalize_commit(...)`
- `fn into_metrics(...) -> Vec<ThresholdMetrics>`

This reduces signature to 5 parameters and eliminates implicit index-alignment invariants.

---
*Reviewed by: Architecture, Complexity*

**Comment by dean0x** [comment]
**`compute_actual_sets` allocates O(K²) HashSets per commit (HIGH CONFIDENCE: 82%)**

Location: `crates/rskim-bench/src/cochange/validate.rs:681`

For each query in `known_ids`, builds K HashSet<FileId> values of size K-1 each. For 500-file commits: 500 sets × 499 elements = ~2 MB overhead. Pattern repeats identically for every commit without reuse.

**Fix:** Pre-compute one `HashSet<FileId>` of all known_ids, then compute intersection counts inline using the full set minus self, matching the `predicted_scratch` reuse pattern.

---
*Reviewed by: Reliability, Performance*

**Comment by dean0x** [comment]
**`chrono_now` hand-rolled Gregorian arithmetic is fragile and untestable (HIGH CONFIDENCE: 82%)**

Location: `crates/rskim-bench/src/bin/cochange_validate.rs:223`

The Hinnant civil_from_days algorithm is implemented inline (~60 lines) to avoid a dependency, but has no unit tests for edge cases like leap-year boundaries (2000-02-29, 2100-03-01). The single test only validates format at current system time, not correctness across calendar boundaries.

**Fix:** Either:
1. Extract pure function `epoch_to_iso8601(secs: u64) -> String` with targeted tests for known dates (2000-02-29, century boundaries), or
2. Use `time` crate's `OffsetDateTime::now_utc()` (already in Cargo.lock via transitive deps) to eliminate hand-rolled logic.

---
*Reviewed by: Rust, Reliability*

**Comment by dean0x** [comment]
**Stale Cargo.lock: phantom `libc` entry for rskim-bench (HIGH CONFIDENCE: 92%)**

Location: `Cargo.lock:2316`

The committed `Cargo.lock` lists `libc` as a direct dependency of `rskim-bench`, but `crates/rskim-bench/Cargo.toml` does not declare it. A prior resolution correctly removed `libc` from Cargo.toml but the lockfile was not regenerated. CI workflows relying on `--locked` to detect drift will not catch real lockfile issues while this phantom entry exists.

**Fix:** Run `cargo generate-lockfile` or `cargo update -w` to regenerate the lockfile, then commit the updated `Cargo.lock`.

This applies ADR-001 (fix noticed issues immediately).

---
*Reviewed by: Dependencies*

**Comment by dean0x** [comment]
## Summary: Lower-Confidence Suggestions (60-79%)

The following issues were identified but fall below the 80% confidence threshold for inline comments. They are worth considering for future improvement:

### Medium Confidence (70-79%)

**Triple intersection walk in sweep_thresholds** (82% Confidence)
- Location: `crates/rskim-bench/src/cochange/validate.rs:730-738`
- For each (threshold, query) pair, precision/recall helpers and TP accumulation each call `predicted.intersection(actual).count()`, walking the same two sets three times per query.
- Fix: Compute intersection count once, derive precision/recall/TP from it.

**Unbounded commit parse phase** (82% Confidence)
- Location: `crates/rskim-bench/src/cochange/validate.rs:558-560`
- Parsing entire git history into memory loads all commit metadata at once. Peak memory per repo could be significant for large repos (10K+ commits).
- Fix: Consider adding `MAX_COMMITS` constant (e.g., 500K) to bail early if repo history unexpectedly large.

**Untestable calendar arithmetic for timestamps** (68% Confidence)
- The Hinnant algorithm is correct but lacks unit tests for leap-year/century boundaries.
- Consider extracting to pure function with targeted tests for known dates.

### Low-Medium Confidence (62-69%)

**Redundant code duplication in clone.rs** (85% Confidence on pre-existing code)
- Location: `crates/rskim-research/src/clone.rs:74-119, 128-169`
- `git_run_with_timeout` and `git_output_with_timeout` share ~40 lines of timeout/SIGKILL/join logic.
- Fix: Extract generic `run_with_timeout<F, T>(cmd, label, timeout_secs, wait_fn: F)` helper.
- Note: This is pre-existing code, but worth addressing while the codebase is fresh.

---

**Comments created on 5 HIGH-confidence issues (80%+)**
**Rate limit status:** 4,992/5,000 remaining

*All findings reviewed by: Architecture, Performance, Reliability, Rust, Dependencies*
